#Acorns of code: The role of bioinformatics in determining the consequences of climate change for cork oak populations.

##What if bioinformatics could help us understand</br>how cork oak tackles climate change?

###Francisco Pina-Martins ([@FPinaMartins](https://twitter.com/FPinaMartins))
####Octávio S. Paulo
#####[https://stuntspt.github.io/BODs2017/](https://stuntspt.github.io/BODs2017/)

<img src="assets/Cover_slide.png" />


.fx: titleslide transparentheading coverslide

---

#The cork oak

* "Analyses of the genomic variation to study cork oak evolution and adaptation: from past to future climatic changes."

![sobreiro](assets/sobreiro.jpg)

.fx: alt_background

---

#The cork oak

* "Analyses of the genomic variation to study cork oak evolution and adaptation: from past to future climatic changes."

![sobreiro](assets/sobreiro.jpg)

* "Programa Doutoral em Biologia e Ecologia das Alterações Globais"

.fx: alt_background

---

#The cork oak

* "Analyses of the genomic variation to study cork oak evolution and adaptation: from past to future climatic changes."

![sobreiro](assets/sobreiro.jpg)

* "Programa Doutoral em Biologia e Ecologia das Alterações Globais"
* Evolutionary biology of *Quercus suber*

.fx: alt_background

---

#The cork oak

* Very broad distribution

.fx: alt_background

---

![hot](assets/hot_summer.jpg)

.fx: imageslide

---

![dist-map](assets/cold_winter.jpg)

.fx: imageslide

---

#The cork oak

* Very broad distribution

![dist-map](assets/dist_map.png)

.fx: alt_background

---

#The cork oak

* Very broad distribution

![dist-map](assets/dist_map_samples.png)

.fx: alt_background

---

#The cork oak

* Plenty of previous studies

.fx: alt_background

---

#The cork oak

* Plenty of previous studies
    * Ecological

.fx: alt_background

---

#The cork oak

* Plenty of previous studies
    * Ecological

![eco-paper](assets/paper_eco.gif)

.fx: compimageslide alt_background

---

#The cork oak

* Plenty of previous studies
    * Ecological
    * Physiological

.fx: alt_background

---

#The cork oak

* Plenty of previous studies
    * Ecological
    * Physiological


![physio-paper](assets/paper_physio.png)

.fx: compimageslide alt_background

---

#The cork oak

* Plenty of previous studies
    * Ecological
    * Physiological
    * Dispersal

.fx: alt_background

---

#The cork oak

* Plenty of previous studies
    * Ecological
    * Physiological
    * Dispersal

![dispersal-paper](assets/paper_disp.png)

.fx: compimageslide alt_background

---

#The cork oak

* Plenty of previous studies
    * Ecological
    * Physiological
    * Dispersal
    * Genetic

.fx: alt_background alt_background

---

#The cork oak

* Plenty of previous studies
    * Ecological
    * Physiological
    * Dispersal
    * Genetic

![genetics-papers](assets/papers_genetics.png)

.fx: compimageslide alt_background

---

#The cork oak

* Hardly any genomic resources

.fx: alt_background

---

#The cork oak

* Hardly any genomic resources
    * No reference genome

.fx: alt_background

---

#The cork oak

* Hardly any genomic resources
    * No reference genome
    * Few GenBank resources (591 "cds" records)

.fx: alt_background

---

#The cork oak

* Hardly any genomic resources
    * No reference genome
    * Few GenBank resources (591 "cds" records)
        * 460 of which are from a paper from our group

.fx: alt_background

---

#The cork oak

* Hardly any genomic resources
    * No reference genome
    * Few GenBank resources (591 "cds" records)
        * 460 of which are from a paper from our group
    * Very few annotations

.fx: alt_background

---

#The strategy

* 454

.fx: alt_background muchsmaller

---

#The strategy

* 454
    * Exploration

.fx: alt_background muchsmaller

---

#The strategy

* 454
    * Exploration
* Sequenom

.fx: alt_background muchsmaller

---

#The strategy

* 454
    * Exploration
* Sequenom
    * Pop. genetics/Evo. history

.fx: alt_background muchsmaller

---

#The strategy

* 454
    * Exploration
* Sequenom
    * Pop. genetics/Evo. history
* GBS

.fx: alt_background muchsmaller

---

#The strategy

* 454
    * Exploration
* Sequenom
    * Pop. genetics/Evo. history
* GBS
    * **G**enome **W**ide **A**ssociation **S**tudy (**GWAS**)

.fx: alt_background muchsmaller

---

#The strategy

* 454
    * Exploration
* Sequenom
    * Pop. genetics/Evo. history
* GBS
    * **G**enome **W**ide **A**ssociation **S**tudy (**GWAS**)

![blue-red-pill](assets/red_pill_blue_pill.jpg)

.fx: alt_background muchsmaller

---

#The strategy

* 454
    * Exploration
* Sequenom
    * Pop. genetics/Evo. history
* GBS
    * **G**enome **W**ide **A**ssociation **S**tudy (**GWAS**)

![blue-red-pill](assets/red_pill_blue_pill.jpg)


<center>**[Take the red pill]**</center>

.fx: muchsmaller

---

#454 data

* No strain information

---

#454 data

* No strain information
* No reference genome

---

#454 data

* No strain information
* No reference genome
* How do we call SNPs?

---

#454 data

* No strain information
* No reference genome
* How do we call SNPs?

![454](assets/454.jpg)

---

#454 data

* Manual methods
* *Custom scripts*
* Develop our own method
* Meet *[4Pipe4](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-016-0892-1)*

<a href="https://github.com/StuntsPT/4Pipe4"><img src="assets/github_green.png" class="forkme" /></a>

.fx: github 

---

![4Pipe4](assets/4pipe4.png)

.fx: imageslide

---

![4Pipe4](assets/4pipe4_2.png)

.fx: imageslide

---

#*4Pipe4*

* Started in 2010
* Written in *python 3*
    * Functional programming
    * Split in modules
    * Small use of classes
* Outperforms every other tested method on **P**ositive **P**redictive **V**alue (**PPV**)
* Published in 2016

<a href="https://github.com/StuntsPT/4Pipe4"><img src="assets/github_green.png" class="forkme" /></a>

.fx: github 

---

#454 data

* Resulted in 402 called SNPs
    * All SNPs were called inside ORFs
    * All had a BLAST match to *something*

* Validation
* Referencing

---

#Sequenom data

* Based on the previous 454 dataset
* 17 sampling sites
* 36 SNPs expanded to 275 individuals
* Little missing data

![Sequenom pic](assets/sequenom.jpg)

---

#Sequenom data analyses

* Fixed dataset
* Performed plenty of test runs
* Great as 'hands on' approach to SNP data analyses

![armory](assets/armory.png)

* **Mostly** manual

---

#Structure_threader

* Automates & paralelizes STRUCTURE & fastStructure analyses

![smith clones](assets/smithclones.jpg)

* Saved a reasonale ammount of analyses time
* Less human intervention == less random error

<a href="https://github.com/StuntsPT/Structure_threader"><img src="assets/github_green.png" class="forkme" /></a>

.fx: github

---

#Structure_threader

* Written in *python 3*
    * Uses <font color="lightgreen">`multiprocessing`</font> module
    * Concurrency, race conditions, multi-threaded I/O
    * Unit tests
    * Uses setup.py

Easy to install from pypi:

    !python
    pip3 install structure_threader --user

<a href="https://github.com/StuntsPT/Structure_threader"><img src="assets/github_green.png" class="forkme" /></a>

.fx: github

---

#GBS data

* Somewhat large dataset (23M reads)
* 95 individuals from 17 sampling sites
* *RAW* data

![spoon](assets/spoon.jpg)

---

#GBS data analyses

* **R**educed **R**epresentation **L**ibraries (**RRLs**) analysis techniques are still under active development

![upgrades](assets/upgrades.jpg)

* Changes in the SNP caller influence the entire process
* Implies **a lot** of repetition

---

#GBS data analyses

* Automation

![makefile](assets/makefile.png)

* Has the nice side effect of making the entire process reproducible
* Can be relatively flexible

---

#RONA

* **R**isk **o**f **N**on-**A**daptedness (**RONA**)

---

#RONA

* **R**isk **o**f **N**on-**A**daptedness (**RONA**)
* Rellstab *et* al. 2016

![RONA_paper](assets/Rellstab.png)

.fx: compimageslide

---

#RONA

* **R**isk **o**f **N**on-**A**daptedness (**RONA**)
* Rellstab *et* al. 2016
* Uses genotype-environment analyses to predict the RONA
* However, there was no implementation
    * Method is simply "described"
    * Refers several time to "in house scripts"
* Despite this "shortcoming", it's still awesome

---

#RONA

* Python to the rescue again

![rescued](assets/rescue.png)

* Implemented as [pyRona](https://github.com/StuntsPT/pyRona)
* Slightly different from the original
* Still WIP

<a href="https://github.com/StuntsPT/pyRona"><img src="assets/github_green.png" class="forkme" /></a>

.fx: github

---

#Biological results

* The short version

![telephone](assets/Telephone.png)


---

#454 & Sequenom

* Found some interesting associations between well known genes and environmental variables
    * eg. "Heat shock protein" && Max. Temperature
    * May provide interesting follow ups
* First evidence that nuclear SNPs and cpDNA tell a different evolutionary history

![STRUCTURE plot](assets/sequenom_str.png)

.fx: alt_background

---
